ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -259,7 +258,8 @@ module Ldap = struct
 
   let query_sid ~name ~kdc ~domain_netbios =
     let key = "objectSid" in
-    let env = env_of_lookup domain_netbios in
+    let env = env_of_krb5 domain_netbios in
+    let name = String.escaped name in
```

Comment:
Escaping must be done by a function that understands LDAP query syntax and escapes characters with special meaning in an LDAP query. `String.escaped` in OCaml just escapes OCaml specific characters.
See here for a list of special characters and how to escape them for LDAP queries: https://docs.microsoft.com/en-gb/windows/win32/adsi/search-filter-syntax?redirectedfrom=MSDN#special-characters

Also depending on what the string argument and where it is used you may have to escape it differently, for example DNs used alone need additional escaping: https://ldapwiki.com/wiki/DN%20Escape%20Values

Reply:
This might be well placed into an in-line sub module where you construct the query:
```
module LDAPquery : sig
  type t
  val sid: string -> string -> t
  val to_string: t -> string
end
```

Internally, `t` would be a sum type that represents the different queries that you want to support. Escaping would be either implemented in `sid` or `to_string` - both would be aware of the role of each parameter.

Reply:
we already have a ldap module and I have put it there, please have a check, thanks @lindig @edwintorok